File: day:64/palindrome_partition.py
from typing import List


def build_partition_table(s):
    matrix = []
    for char in s:
        matrix.append([0] * len(s))
    # all single char are palindrome themselves
    for i in range(len(s)):
        matrix[i][i] = 1
    # check palindrome for substrings (length of 2 and above)
    for col in range(1, len(s)):
        for row in range(col):
            if row + 1 == col:
                # two chars case
                if s[row] == s[col]:
                   matrix[row][col] = 1
            else:
                # more than two chars
                # check if the bottom-left is 1
                if matrix[row + 1][col - 1]:
                    # check whether the leftmost/rightmost chars are same
                    # substring between them is palindrome
                    if s[row] == s[col]:
                        matrix[row][col] = 1
    return matrix

class Solution:
    def __init__(self):
        self.pMatrix = []
        self.result = []

    def dfs(self, start, s, current_list):
        if start >= len(s):
            self.result.append(current_list[:])
        else:
            for end in range(start, len(s)):
                # check the substring (start, end)
                if self.pMatrix[start][end]:
                    current_list.append(s[start:end+1])
                    self.dfs(end + 1, s, current_list)
                    # backtrack it
                    current_list.pop()

    def partition(self, s: str) -> List[List[str]]:
        self.pMatrix = build_partition_table(s)
        self.dfs(0, s, [])
        r = self.result[:]
        result = []
        return r
        

File: day:64/palindromePartition.java
class Solution {
    List<List<String>> result = new ArrayList<List<String>>();
    public int[][] buildPTable(String s) {
        int pMatrix[][] = new int[s.length()][s.length()];
        for(int j = 0; j < s.length(); j++) {
            for(int i = 0; i < s.length(); i++) {
                pMatrix[i] = new int[s.length()];
            }
        }
        // all single char are palindrome themselves
        for(int i = 0; i < s.length(); i++) {
            pMatrix[i][i] = 1;
        }
        // check palindrome for substrings (length of 2 and above)
        for(int col = 1; col < s.length(); col++) {
            for(int row = 0; row < col; row++) {
                if(row + 1 == col) {
                    // two chars case
                    if(s.charAt(row) == s.charAt(col)) {
                        pMatrix[row][col] = 1;
                    }
                }
                else {
                    // more than two chars
                    // check if the bottom-left is 1
                    if(pMatrix[row+1][col-1] == 1) {
                        // check whether the leftmost/rightmost chars are same;
                        // substring between them is palindrome
                        if(s.charAt(row) == s.charAt(col)) {
                            pMatrix[row][col] = 1;
                        }
                    }
                }
            }
        }
        return pMatrix;
    }
    public void dfs(int start, String str, List<String> currentList, int[][] pMatrix) {
        if(start >= str.length()) {
            result.add(new ArrayList<String>(currentList));
        }
        else {
            for(int end = start; end < str.length(); end++) {
                // check the substring (start, end)
                if(pMatrix[start][end] == 1) {
                    currentList.add(str.substring(start, end + 1));
                    dfs(end + 1, str, currentList, pMatrix);
                    // backtrack it
                    currentList.remove(currentList.size() - 1);
                }
            }
        }
    }
    public List<List<String>> partition(String s) {
        int[][] matrix = buildPTable(s);
        dfs(0, s, new ArrayList<String>(), matrix);
        return result;
    }
}
